### PR TITLE
Stable version

### PR DESCRIPTION
**Pull request by IlyaHrushhyov**

### CODE REVIEW

File: Catering.BusinessLogicLayer/Services/FeedBackServices.cs [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        public FeedBackServices(IRepositoryManager repositoryManager): base(repositoryManager)
+        {
+            
+        }
+
+        public async Task<IEnumerable<Feedback>> GetFeedbacksAsync() => 
+            await RepositoryManager.Feedback.GetFeedbacksAsync(trackChanges:false);
+
+        public async Task<IEnumerable<Feedback>> GetUserFeedbacksAsync(string userId) =>
+            await RepositoryManager.Feedback.GetUserFeedbacksAsync(userId, trackChanges:false);
```
Comment (DmitryRyder) [inline Catering.BusinessLogicLayer/Services/FeedBackServices.cs:22]: unnecessary naming parameter - trackChanges

File: Catering.BusinessLogicLayer/Services/FeedBackServices.cs [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+{
+    public class FeedBackServices:ServicesBase, IFeedbackServices
+    {
+        public FeedBackServices(IRepositoryManager repositoryManager): base(repositoryManager)
+        {
+            
+        }
+
+        public async Task<IEnumerable<Feedback>> GetFeedbacksAsync() => 
+            await RepositoryManager.Feedback.GetFeedbacksAsync(trackChanges:false);
```
Comment (DmitryRyder) [inline Catering.BusinessLogicLayer/Services/FeedBackServices.cs:19]: unnecessary naming parameter - trackChanges

File: Contracts/Catering.Contracts.csproj [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Contracts/Catering.Contracts.csproj]: All libraries have been moved to .Net 5

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs]: All unnecessary usings have been removed

File: Catering.WebApi/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        }
+
+        public IConfiguration Configuration { get; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.ConfigureDataAccess(Configuration.GetConnectionString("sqlConnection"));
+
+            services.AddAutoMapper(typeof(Catering.BusinessLogicLayer.MappingProfile));
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Startup.cs]: Full path replaced by "using Catering.BusinessLogicLayer"

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    [ApiController]
+    public class AuthenticationController : ControllerBase
+    {
+        
+        private readonly ILoggerManager _logger;
+        private readonly IMapper _mapper;
+        private readonly UserManager<User> _userManager;
+        private readonly IAuthenticationManager _authManager;
+        public AuthenticationController(ILoggerManager logger, IMapper mapper,
+       UserManager<User> userManager, IAuthenticationManager authManager)
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs:25]: Solved by adding stylecop

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+            
+            _authManager = authManager;
+        }
+
+        [HttpPost]
+        [ServiceFilter(typeof(ValidationFilterAttribute))]
+        public async Task<IActionResult> RegisterUser([FromBody] UserForRegistrationDto userForRegistration)
+        {
+           
+
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs]: Solved by adding stylecop

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            var user = _mapper.Map<User>(userForRegistration);
+            var result = await _userManager.CreateAsync(user, userForRegistration.Password);
+            
+            if (!result.Succeeded)
+            {
+                foreach (var error in result.Errors)
+                {
+                    ModelState.TryAddModelError(error.Code, error.Description);
+                }
+                return BadRequest(ModelState);
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs]: Solved by adding stylecop

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            }
+            await _userManager.AddToRolesAsync(user, userForRegistration.Roles);
+            return StatusCode(201);
+        }
+
+        [HttpGet, Authorize(Roles = "Manager")]
+        public async Task<IActionResult> GetUsers()
+        {
+            var users = await _userManager.Users.ToListAsync();
+
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs]: Solved by adding stylecop

File: Catering.WebApi/ServiceExtensions/AuthenticationManager.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Security.Claims;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Catering.WebApi
+{
+    public class AuthenticationManager:IAuthenticationManager
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/ServiceExtensions/AuthenticationManager.cs]: Solved by adding stylecop

File: Catering.WebApi/ServiceExtensions/AuthenticationManager.cs [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        public AuthenticationManager(UserManager<User> userManager, IConfiguration
+       configuration)
+        {
+            _userManager = userManager;
+            _configuration = configuration;
+        }
+        public async Task<bool> ValidateUser(UserForAuthenticationDto userForAuth)
+        {
+            _user = await _userManager.FindByNameAsync(userForAuth.UserName);
+            return (_user != null && await _userManager.CheckPasswordAsync(_user,
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/ServiceExtensions/AuthenticationManager.cs]: Fixed by adding stylecop

File: Catering.WebApi/ServiceExtensions/AuthenticationManager.cs [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+           userForAuth.Password));
+        }
+        public async Task<string> CreateToken()
+        {
+            var signingCredentials = GetSigningCredentials();
+            var claims = await GetClaims();
+            var tokenOptions = GenerateTokenOptions(signingCredentials, claims);
+            return new JwtSecurityTokenHandler().WriteToken(tokenOptions);
+        }
+        private SigningCredentials GetSigningCredentials()
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/ServiceExtensions/AuthenticationManager.cs:46]: Fixed by adding stylecop

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+       UserManager<User> userManager, IAuthenticationManager authManager)
+        {
+            _logger = logger;
+            _mapper = mapper;
+            _userManager = userManager;
+            
+            _authManager = authManager;
+        }
+
+        [HttpPost]
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs]: Added name of route for each controller method

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+           
+
+            var user = _mapper.Map<User>(userForRegistration);
+            var result = await _userManager.CreateAsync(user, userForRegistration.Password);
+            
+            if (!result.Succeeded)
+            {
+                foreach (var error in result.Errors)
+                {
+                    ModelState.TryAddModelError(error.Code, error.Description);
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs]: Replaced by `return Unauthorized(result.ToString())`

File: Catering.WebApi/Controllers/AuthenticationController.cs [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        [HttpPost]
+        [ServiceFilter(typeof(ValidationFilterAttribute))]
+        public async Task<IActionResult> RegisterUser([FromBody] UserForRegistrationDto userForRegistration)
+        {
+           
+
+            var user = _mapper.Map<User>(userForRegistration);
+            var result = await _userManager.CreateAsync(user, userForRegistration.Password);
+            
+            if (!result.Succeeded)
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/AuthenticationController.cs:40]: Replaced by `return Unauthorized(result.ToString())`

File: Catering.WebApi/Controllers/MenuItemController.cs [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+
+            var category = await _categoryServices.GetCategoryAsync(categoryId, trackChanges: false);
+            if (category == null)
+            {
+                _logger.LogInfo($"Category with id: {categoryId} doesn't exist in the database.");
+                return NotFound();
+            }
+            var menuItemEntity = _mapper.Map<MenuItem>(menuItem);
+
+            await _menuItemServices.CreateMenuItemAsync(categoryId, menuItemEntity);
```
[the remark this answers is not in the document]
Comment (IlyaHrushhyov) [reply Catering.WebApi/Controllers/MenuItemController.cs:88]: I found that i already checked if added entity exists before call `CreateMenuItemAsync()`, so i removed unnecessary logic from `CreateMenuItemAsync()` and added `return NotFound($"Category with id: {categoryId} doesn't exist in the database.")` to `MenuItemController`.
Such changes have been applied to the following services and controllers:
+ OrderedItemService and OrderedItemController
+ MenuItemService and MenuItemController

File: Catering.BusinessLogicLayer/Services/FeedBackServices.cs (hunk 1 above)
Comment (IlyaHrushhyov) [reply Catering.BusinessLogicLayer/Services/FeedBackServices.cs]: Removed trackChanges naming parameter

File: Catering.BusinessLogicLayer/Services/FeedBackServices.cs (hunk 2 above)
Comment (IlyaHrushhyov) [reply Catering.BusinessLogicLayer/Services/FeedBackServices.cs]: Removed  naming parameter - trackChanges

**Comment by DmitryRyder** [comment]
Pull request should be merged

**Review by DmitryRyder — APPROVED** [review]

**Pull request merged**